sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1658,6 +1659,43 @@ class Dataset[T] private[sql](
   def groupByKey[K](func: MapFunction[T, K], encoder: Encoder[K]): KeyValueGroupedDataset[K, T] =
     groupByKey(func.call(_))(encoder)
 
+
+  /**
+   * Aggregates the elements of this Dataset in a multi-level tree pattern.
+   *
+   * @param depth suggested depth of the tree (default: 2)
+   */
+  private[spark] def treeAggregate[U : Encoder : ClassTag](zeroValue: U)(
+      seqOp: (U, T) => U,
+      combOp: (U, U) => U,
+      depth: Int = 2): U = {
+    require(depth >= 1, s"Depth must be greater than or equal to 1 but got $depth.")
```

Comment:
why would depth 1 make sense?

Reply:
This is copied from `rdd.treeAggregate`. Actually it does aggregation at each partition, then do a folding. It is almost equal to `rdd.aggregate`. Logically depth 1 is reasonable, although it is not useful actually, because you can call `aggregate` if you want depth 1.